rfcs/0182-user-management.md

```diff
@@ -0,0 +1,306 @@
+---
+feature: I'm Gonna Build My Own Home Tool With Blackjack and Modules!
+start-date: 2024-09-20
+author: Anderson Torres
+co-authors: @nyabinary
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Enhance Nixpkgs monorepo with a declarative management system for basic users.
+
+# Terminology
+[terminology]: #terminology
+
+Henceforth,
+
+- The typical unprivileged user of a system will be called _basic user_;
+- The typical privileged user of a system will be called _superuser_;
+
+# Motivation
+[motivation]: #motivation
+
+[Nixpkgs](https://github.com/NixOS/nixpkgs) is by a far margin the largest
+packageset in this existence, according to
+[Repology](https://repology.org/repository/nix_unstable).
+
+In principle, currently the Nix interpreter and evaluator already leverage
+Nixpkgs for basic users. However, this raw usage is not very ergonomical,
+especially when compared to the structured model provided by NixOS.
+
+In this panorama, the average user has two extreme options here:
+
+- system-wide configuration via NixOS, requiring superuser privileges;
+- ad-hoc management via raw Nix commands, especially the not-recommended
+  `nix-env`.
+
+This RFC proposes to fill this gap: a structured model of system management for
+basic users that does not require superuser privileges to be deployed.
+
+# Detailed Design
+[design]: #detailed-design
+
+In a dedicated directory inside the Nixpkgs monorepo a well-crafted set of core
+modules, plus some companion modules, will be created and maintained.
+
+Further, a companion toolset (henceforth called `hometool`) to manipulate and
+deploy the home environment will be provided.
+
+This toolset plus module set should provide the following properties:
+
+- Declarativeness
+
+  Users can specify the configuration of their systems in Nix language, by a set
+  of detailed descriptions.
+
+- Immutability
+
+  As a consequence of declarativeness, the resulting descriptions are immutable.
+  It allows comparing them by knowing precisely what changed between
+  configurations.
+
+- Customizability
+
+  Users can derive specialized module configurations from current module set.
```

Comment:
Exempli gratia, the user should be able to override the `emacs` program from a `programs.emacs` module.

(This is a recent concern in NixOS modules; many of them hardcode the main program.)

Reply:
That doesn't sound much like a "specialized module configuration" to me, just ensuring support for what should be basic options in a module. The wording here implies extending/editing the base modules to fit individual needs; simply introducing a `package` option fails to do this IMO

Reply:
> The wording here implies extending/editing the base modules to fit individual needs

This would be interesting too.

Reply:
That'd be awesome even in NixOS honestly, it'd be nice to have a _slightly_ nicer interface to extend modules, particularly the ones that use submodule types (sure its not hard, but it could be made simpler maybe)